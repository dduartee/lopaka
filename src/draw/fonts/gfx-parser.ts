const GFX_BITMAP_REGEX = /const uint8_t\s+(\w+)\[\]\s+PROGMEM\s+=\s+\{([^}]+)\}/;
const GFX_GLYPHS_REGEX = /const GFXglyph\s+(\w+)\[\]\s+PROGMEM\s+=\s+{([^;]+)};/;
const GFX_FONT_REGEX = /const GFXfont\s+(\w+)\s+PROGMEM\s+=\s+{([^}]+)}/;

function readBit(byte: number, bit: number) {
    return (byte & (1 << bit)) >> bit;
}

function writeBit(byte: number, bit: number, value: number) {
    return byte | (value << bit);
}

export default function parseGFX(source: string): FontPack {
    // clean up comments
    source = source.replace(/\/\/.*/g, '');
    // parse bitmap
    const bitmapMatch = source.match(GFX_BITMAP_REGEX);
    if (!bitmapMatch) {
        throw new Error('No bitmap data found');
    }
    const bitmap = bitmapMatch[2].split(',').map((v) => parseInt(v, 16));
    const glyphsMatch = source.match(GFX_GLYPHS_REGEX);
    const rawGlyphs = glyphsMatch[2]
        .replace(/\s+/g, '')
        .split('},')
        .filter((g) => g !== '')
        .map((g) => {
            const values = g.replace(/{/g, '').split(',');
            return values.map((value) => {
                const parsed = Number(value.replace('}', ''));
                return isNaN(parsed) ? 0 : parsed;
            });
        });
    const fontMatch = source.match(GFX_FONT_REGEX);
    let fontData = fontMatch[2].split(',').map((v) => parseInt(v.trim(), 16));
    const first = fontData[2];
    fontData = null;
    const glyphs = new Map<number, FontGlyph>();
    rawGlyphs.forEach((glyph, n) => {
        const [bitmapOffset, width, height, xAdvance, xOffset, yOffset] = glyph;
        const code = n + first;
        const bytes = decodeBitmap(
            bitmap.slice(bitmapOffset, bitmapOffset + 1 + Math.ceil(width * height) / 8),
            width,
            height
        );
        glyphs.set(code, {
            code,
            bytes,
            bounds: [xOffset, -yOffset, width, height],
            scalableSize: [width, height],
            deviceSize: [width, height],
            xAdvance,
        });
    });
    const maxWidth = Math.max(...rawGlyphs.map((g) => g[1] + g[3] + g[4]));
    const maxHeight = Math.max(...rawGlyphs.map((g) => g[2] - (g[5] + g[2])));
    const font = {
        meta: {
            name: fontMatch[1],
            size: {
                points: maxHeight,
                resolutionX: maxHeight * 10,
                resolutionY: maxHeight * 10,
            },
            bounds: [0, 0, maxWidth, maxHeight],
            properties: {
                fontDescent: 0,
                fontAscent: 0,
                defaultChar: first,
            },
            totalChars: glyphs.size,
        },
        glyphs,
    };
    return font;
}

export function decodeBitmap(gfxBytes: number[], width: number, height: number) {
    const bdfBytes = [];
    let byte = 0;
    let outBitPos = 0;
    for (let n = 0; n < width * height; n++) {
        let bytePos = Math.floor(n / 8);
        let bitPos = 7 - (n % 8);
        byte = writeBit(byte, 7 - (outBitPos % 8), readBit(gfxBytes[bytePos], bitPos));
        outBitPos++;
        if (outBitPos === width) {
            bdfBytes.push(byte);
            byte = 0;
            outBitPos = 0;
        } else if (outBitPos > 1 && outBitPos % 8 === 0) {
            bdfBytes.push(byte);
            byte = 0;
        }
    }
    return bdfBytes;
}

export function encodeBitmap(bdfBytes: number[], width: number): number[] {
    const gfxBytes: number[] = [];
    let byte = 0;
    let bitPos = 0;
    const bytesPerRow = Math.ceil(width / 8);

    for (let row = 0; row < bdfBytes.length; row += bytesPerRow) {
        for (let byteIndex = 0; byteIndex < bytesPerRow; byteIndex++) {
            const currentByte = bdfBytes[row + byteIndex] || 0; // Handle missing bytes gracefully
            for (let bitIndex = 0; bitIndex < 8; bitIndex++) {
                const col = byteIndex * 8 + bitIndex;
                if (col >= width) break; // Avoid exceeding the glyph width

                const bdfBit = readBit(currentByte, 7 - bitIndex);
                byte = writeBit(byte, 7 - bitPos, bdfBit);
                bitPos++;

                if (bitPos === 8) {
                    gfxBytes.push(byte);
                    byte = 0;
                    bitPos = 0;
                }
            }
        }
    }

    if (bitPos > 0) {
        gfxBytes.push(byte);
    }

    return gfxBytes;
}
