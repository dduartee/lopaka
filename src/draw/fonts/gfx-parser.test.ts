import {describe, expect, test} from 'vitest';
import {decodeBitmap, encodeBitmap} from './gfx-parser';

function printBytes(bytes) {
    let str = '';
    for (let i = 0; i < bytes.length; i++) {
        const bits = bytes[i].toString(2);
        str += ('0'.repeat(8 - bits.length) + bits).replaceAll('1', '▪').replaceAll('0', '▫') + '\n';
    }
    return str;
}

describe('gfx-parser', () => {
    test('parseGFX', () => {
        const data = [0xf3, 0x1c, 0x53, 0x54, 0xe5, 0x48];
        const width = 3;
        const height = 5;
        for (let i = 0; i < 3; i++) {
            expect(printBytes(decodeBitmap(data.slice(i * 2, i * 2 + 2), width, height))).toMatchSnapshot();
        }
    });

    test('decode gfx bitmap', () => {
        const gfxBitmap = [0x49, 0x24, 0x92, 0x48, 0x01, 0xf8];
        const bdfBitmap = [0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x00, 0x00, 0x00, 0xe0, 0xe0];
        const width = 3;
        const height = 15;
        const decoded = decodeBitmap(gfxBitmap, width, height);
        expect(decoded).toEqual(bdfBitmap);
    });

    test('encode gfx bitmap', () => {
        const gfxBitmap = [0x49, 0x24, 0x92, 0x48, 0x01, 0xf8];
        const bdfBitmap = [0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x40, 0x00, 0x00, 0x00, 0xe0, 0xe0];
        const width = 3;
        const height = 15;
        const encoded = encodeBitmap(bdfBitmap, width, height);
        expect(encoded).toEqual(gfxBitmap);
    });
});
