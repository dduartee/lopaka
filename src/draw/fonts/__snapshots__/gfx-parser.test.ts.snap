// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`gfx-parser > parseGFX 1`] = `
"▪▪▪▫▫▫▫▫
▪▫▫▫▫▫▫▫
▪▪▫▫▫▫▫▫
▫▫▪▫▫▫▫▫
▪▪▫▫▫▫▫▫
"
`;

exports[`gfx-parser > parseGFX 2`] = `
"▫▪▫▫▫▫▫▫
▪▫▫▫▫▫▫▫
▪▪▫▫▫▫▫▫
▪▫▪▫▫▫▫▫
▫▪▫▫▫▫▫▫
"
`;

exports[`gfx-parser > parseGFX 3`] = `
"▪▪▪▫▫▫▫▫
▫▫▪▫▫▫▫▫
▫▪▫▫▫▫▫▫
▪▫▫▫▫▫▫▫
▪▫▫▫▫▫▫▫
"
`;
